driver/pelican/types/thermSchedule.go

```diff
@@ -0,0 +1,246 @@
+package types
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/parnurzeal/gorequest"
+	rrule "github.com/teambition/rrule-go"
+)
+
+// Login, Authentication, Thermostat ID Retrieval Structs
+type thermIDRequest struct {
+	Resources []thermIDResources `json:"resources"`
+}
+
+type thermIDResources struct {
+	Children    []thermIDChild `json:"children"`
+	GroupId     string         `json:"groupId"`
+	Permissions string         `json:"permissions"`
+}
+
+type thermIDChild struct {
+	Id          string `json:"id"`
+	Permissions string `json:"permissions"`
+}
+
+// Thermostat Settings Structs
+type settingsRequest struct {
+	Epnum    float64         `json:"epnum"`
+	Id       string          `json:"id"`
+	Nodename string          `json:"nodename"`
+	Userdata settingsWrapper `json:"userdata"`
+}
+
+type settingsWrapper struct {
+	Epnum    float64 `json:"epnum"`
+	Fan      string  `json:"fan"`
+	Nodename string  `json:"nodename"`
+	Repeat   string  `json:"repeat"`
+}
+
+// Thermostat Schedule By Day Decoding Structs
+type scheduleRequest struct {
+	ClientData scheduleSetTimes `json:"clientdata"`
+}
+
+type scheduleSetTimes struct {
+	SetTimes []scheduleTimeBlock `json:"setTimes"`
+}
+
+type scheduleTimeBlock struct {
+	HeatSetting float64 `json:"heatSetting"`
+	CoolSetting float64 `json:"coolSetting"`
+	StartValue  string  `json:"startValue"`
+	System      string  `json:"systemDisplay"`
+}
+
+// Thermostat Schedule Structs
+
+// Struct mapping each day of the week to its daily schedule
+type ThermostatSchedule struct {
+	DaySchedules map[string]([]ThermostatBlockSchedule) `msgpack:"day_schedules"`
+}
+
+// Struct containing data defining the settings of each schedule block
+type ThermostatBlockSchedule struct {
+	CoolSetting float64 `msgpack:"cool_setting"`
+	HeatSetting float64 `msgpack:"heat_setting"`
+	System      string  `msgpack:"system"`
+	Time        string  `msgpack:"time"`
+}
+
+// Time Constant for Cookie Refresh. 720 Hours = 30 Days
+const cookieDuration = time.Duration(720)
+
+var lastCookie = time.Now()
```

Comment:
We definitely want to avoid using a global variable like this. Remember that we may have multiple Pelican instances, and they would all share this variable, which is bad.

The alternative I thought of is to add another field to the Pelican struct for this timestamp. Thoughts?

Reply:
Good point, that would not be good at all. In my latest edit I added a field called "cookieTime" which would track when the cookie was created.

It's initialized to a value of nil. I changed the conditional to refresh the cookie if 1. the cookie field was nil or 2. if the time since cookie time was greater than the duration. I reset the "cookie" and "cookieTime" fields within the "setCookieAndID" method.

I think adding a field to the Pelican struct makes the most sense since each Pelican will have to have its own "timestamp" of sorts for its cookie.

Reply:
Follow up: I made the value of the "cookieTime" field a value, not a pointer, per the go documentation suggestion [here](https://golang.org/pkg/time/#Time) where it states "Programs using times should typically store and pass them as values, not pointers. That is, time variables and struct fields should be of type time.Time, not *time.Time."